### PR TITLE
Migrating from coherence to guardian

### PR DESCRIPTION
**Pull request by ghatighorias**
TO-DOs
- [x] sent the password reset token to the user

- [x] write tests for session_controller

- [x] write tests for password_controller

- [x] add coverage for user schema

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -48,3 +50,10 @@ config :email_checker,
   validations: [EmailChecker.Check.Format]
 
 import_config "#{Mix.env}.exs"
+
+config :guardian, Guardian,
+   issuer: "CoursePlanner.#{Mix.env}",
+   ttl: {1, :days},
+   verify_issuer: true,
+   serializer: CoursePlanner.Auth.GuardianSerializer,
+   secret_key: to_string(Mix.env) <> "SuPerseCret_aBraCadabrA"
```
[the remark this answers is not in the document]
Comment (ghatighorias) [reply config/config.exs]: yes I saw it in some posts that I should use it like that. but totally for got it. :P

[the remark this answers is not in the document]
Comment (ghatighorias) [reply config/config.exs]: maybe we can fill it by a combination of a default value and server adjustable part so user can simply modify it.

[the remark this answers is not in the document]
Comment (rhnonose) [reply config/config.exs]: I'm not sure if we should let the user change.

File: lib/course_planner/accounts/user.ex [hunk 2]
```diff
@@ -29,36 +35,69 @@ defmodule CoursePlanner.Accounts.User do
...
   end
 
-  def changeset(model, params, :password) do
+  def changeset(model, params, :password_reset) do
     model
     |> cast(params,
       ~w(password password_confirmation reset_password_token reset_password_sent_at))
-    |> validate_coherence_password_reset(params)
+    |> validate_required(:password)
+    |> validate_length(:password, min: 6)
```
[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner/accounts/user.ex]: shall I increase it to 8 ?

[the remark this answers is not in the document]
Comment (rhnonose) [reply lib/course_planner/accounts/user.ex]: Yes

File: lib/course_planner/accounts/user.ex [hunk 3]
```diff
@@ -74,8 +113,59 @@ defmodule CoursePlanner.Accounts.User do
...
-      true -> changeset
-      false -> Changeset.add_error(changeset, field, "has invalid format", [validation: :format])
+         true -> changeset
+         false -> Changeset.add_error(changeset, field, "has invalid format", [validation: :format])
+       end
+  end
+
+  def validate_current_password(%{changes: changes, valid?: true} = changeset) do
+    password_hash = Map.get(changeset.data, :password_hash)
+    current_password = Map.get(changes, :current_password)
```
[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner/accounts/user.ex]: get_field will check both changeset.changes and changeset.data ( if i am right ) but here I needed to explicitly check just one of each

File: lib/course_planner_web/controllers/auth/password_controller.ex [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    user = Repo.get_by(User, email: trimmed_downcased_email)
+
+    if user do
+      updated_user = set_new_password_reset_token(user)
+      password_reset_url =  Helpers.password_url(conn, :edit, updated_user.reset_password_token)
+      UserEmail.send_user_email(:password, updated_user, password_reset_url)
+    end
+
+    conn
+    |> put_flash(:info, "If the email address is registered, an emaill will be send to it")
```
[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner_web/controllers/auth/password_controller.ex]: hahahaha I always sucked at spelling . i think i have dyslexia

File: lib/course_planner/accounts/user.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhnonose) [reply lib/course_planner/accounts/user.ex]: Ah, not I understand what you did there.

[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner/accounts/user.ex]: upgraded to 8

File: lib/course_planner_web/controllers/auth/password_controller.ex (hunk 4 above)
[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner_web/controllers/auth/password_controller.ex]: done

File: lib/course_planner_web/templates/layout/session_layout.html.eex [hunk 5]
```diff
@@ -7,7 +7,7 @@
     <meta name="description" content="">
     <meta name="author" content="">
 
-    <title><%= Coherence.Config.title() %></title>
+    <title>"Course Planner"</title>
```
[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner_web/templates/layout/session_layout.html.eex]: done

File: lib/course_planner/accounts/user.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghatighorias) [reply lib/course_planner/accounts/user.ex]: upgraded to 8

**Comment by ghatighorias** [comment]
@rhnonose 
I have added code for updating "last_sign_in_at" and "failed_attempts" fields
check Users.update_login_fields function and "create" function from CoursePlannerWeb.Auth.SessionController

**Pull request merged**